site/public/ts/shared/utils/Exporter.ts

```diff
@@ -0,0 +1,100 @@
+import {DigitalCircuitView} from "site/digital/views/DigitalCircuitView";
+import {V} from "Vector";
+import {Browser} from "core/utils/Browser";
+
+// Returns a 256x256 canvas on which the given circuit is centered
+//! No idea what happens if the designer is empty, divide-by-zero seems plausible
+//? how empty designers should be handled -- return empty canvas?
+function RenderCircuit(designer: DigitalCircuitDesigner): HTMLCanvasElement {
+    // 256 is arbitrary, could increase or decrease for more detail or less memory respectively
+    const THUMBNAIL_SIZE = 256;
+
+    // Find bounding box of the circuit
+    let xmin = Infinity;
+    let ymin = Infinity;
+    let xmax = -Infinity;
+    let ymax = -Infinity;
+    for (let c of designer.getObjects()) {
+        const cullbox = c.getCullBox();
+        const max = cullbox.getTopRight();
+        const min = cullbox.getBottomLeft();
+        xmin = Math.min(min.x, xmin);
+        ymin = Math.min(min.y, ymin);
+        xmax = Math.max(max.x, xmax);
+        ymax = Math.max(max.y, ymax);
+    }
+
+    //! Warning: several pieces of CircuitView classes rely on there being a screen and a window
+    // I think using this vw and vh cancels it out, but I'm not confident in that
+    const canvas = document.createElement("canvas");
+    const vw = THUMBNAIL_SIZE/window.innerWidth;
+    const vh = THUMBNAIL_SIZE/window.innerHeight;
+    const view = new DigitalCircuitView(canvas, vw, vh);
+
+    // Center and zoom the camera so everything fits with no distortion
+    const center = V((xmin + xmax)/2, (ymin + ymax)/2);
+    const camera = view.getCamera();
+    camera.setPos(center);
+    // Zoom out 2.5% more than we need, just for a little padding
+    const zoom = Math.max((xmax - xmin)/THUMBNAIL_SIZE, (ymax - ymin)/THUMBNAIL_SIZE) * 1.025;
+    camera.zoomBy(zoom);
+
+    // Create background the same color as the site to composite the image on top of
+    // NOTE: can't just fill the first canvas before the render since the first thing the renderer does it clear it
+    const composite = document.createElement("canvas");
+    composite.width = THUMBNAIL_SIZE;
+    composite.height = THUMBNAIL_SIZE;
+    const context = composite.getContext("2d");
+    context.fillStyle = window.getComputedStyle(document.body, null).getPropertyValue("background-color");
+    context.fillRect(0, 0, THUMBNAIL_SIZE, THUMBNAIL_SIZE);
+
+    // Do the render, and composite it on top of the background
+    view.render(designer, [], undefined);
+    context.drawImage(canvas, 0, 0, THUMBNAIL_SIZE, THUMBNAIL_SIZE);
+    return composite;
+}
+
+export function WriteCircuit(designer: DigitalCircuitDesigner, name: string, thumbnail: boolean = false): string {
+    const writer = new XMLWriter(designer.getXMLName());
+    writer.setVersion("1.1");
+    writer.setName(name);
+    if (thumbnail) {
+        // Render to a small temporary canvas (which gets garbage-collected once done)
```

Comment:
`document.createElement` creates the canvas, do we have to `document.removeElement` or something? I'd assume it really doesn't get garbage collected until removed from the DOM strictly. We should also make sure this is true because canvas's use a ton of memory.

Reply:
I was also concerned about this, but eventually assumed it must be garbage collected since the only removal method I found at the time, `canvas.parentNode.removeChild` was crashing. On a second look through, there's another way to do it, just with `canvas.remove()`. Tested, it works.